### PR TITLE
Implement arithmetic overflow changes

### PR DESCRIPTION
**Pull request by Aatch**
Implements most of https://github.com/rust-lang/rfcs/pull/560. Errors that I encountered from the checks during building were fixed.

The checks for division, remainder and bit-shifting have not been implemented yet.

cc @nikomatsakis

### CODE REVIEW

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -547,6 +547,15 @@ extern "rust-intrinsic" {
     pub fn u64_mul_with_overflow(x: u64, y: u64) -> (u64, bool);
 }
 
+#[cfg(not(stage0))]
+extern "rust-intrinsic" {
+    /// Returns (a + b) mod 2^N, where N is the width of N in bits.
+    pub fn overflowing_add<T>(a: T, b: T) -> T;
```
Comment (sfackler) [inline src/libcore/intrinsics.rs]: Any reason these are overflowing_* while the methods are wrapping_*? I think wrapping_* is a bit more descriptive, fwiw.

**Comment by brson** [comment]
Thanks @aatch and @cmr

File: src/test/run-fail/overflowing-mul.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// error-pattern:thread '<main>' panicked at 'arithmatic operation overflowed'
+
+fn main() {
+    let x = 200u8 + 4u8;
```
Comment (gereeter) [inline src/test/run-fail/overflowing-mul.rs]: Should this be multiplication?

Comment (Aatch) [inline src/test/run-fail/overflowing-mul.rs]: Probably, yes. Blame copy-paste. (And @cmr, since it's his code :P)

**Comment by nikomatsakis** [comment]
This looks overall really nice. I'll give it a more detailed look soon.

**Comment by nikomatsakis** [comment]
Argh, slipped off my radar! I'm sorry about that. Tomorrow. (Please feel free to ping me.)

File: src/libcore/intrinsics.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/libcore/intrinsics.rs]: Yes, we ought to give them consistent names.

File: src/libcore/num/wrapping.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+use ops::*;
+
+#[cfg(not(stage0))]
+use intrinsics::{overflowing_add, overflowing_sub, overflowing_mul};
+
+pub trait WrappingOps {
+    fn wrapping_add(self, rhs: Self) -> Self;
+    fn wrapping_sub(self, rhs: Self) -> Self;
+    fn wrapping_mul(self, rhs: Self) -> Self;
```
Comment (nikomatsakis) [inline src/libcore/num/wrapping.rs]: What about `wrapping_div` and `wrapping_mod`? (same with the intrinsics)

File: src/libcore/num/wrapping.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        Wrapping(self.0 & other.0)
+    }
+}
+
+impl<T:WrappingOps+Shl<uint,Output=T>> Shl<uint> for Wrapping<T> {
+    type Output = Wrapping<T>;
+
+    #[inline(always)]
+    fn shl(self, other: uint) -> Wrapping<T> {
+        Wrapping(self.0 << other)
```
Comment (nikomatsakis) [inline src/libcore/num/wrapping.rs]: Similarly, shifts have extra overflow checks defined for them.

File: src/libcore/num/wrapping.rs (hunk 3 above)
Comment (nagisa) [inline src/libcore/num/wrapping.rs]: These two methods only make sense for `f32

**Comment by steveklabnik** [comment]
This doesn't currently merge.

**Comment by nikomatsakis** [comment]
Now that the RFC has been accepted, we can rebase, updated, and land this PR. @Aatch are you stll interested in pursuing that? Can you talk to me on IRC (I've looked for you a few times but not found you, bad timing I guess)?

**Comment by mahkoh** [comment]
This patch doesn't address atomic operations.

**Comment by mahkoh** [comment]
More precisely: If overflow caused by fetch_add or fetch_sub is detected the process must be aborted. Panic is not possible because other threads might already have consumed the value.

**Comment by nikomatsakis** [comment]
@mahkoh and I talked briefly about atomics on IRC. The options seem to be:

1. Just wrap. Document it, perhaps change name of methods.
2. Abort process as @mahkoh suggests.
3. Abort the thread, although the observed value may have been seen elsewhere.
4. Wrap but return some indication of when wrapping occurred.
5. Fallback to a CAS loop.

Each has its own downsides. The first option is obviously prone to programmer error, particularly if someone is naively using an atomic to achieve some kind of global counter. That said, atomics are fairly specialized, and perhaps it should be expected. (Note: If we *don't* wrap by default, then clearly we need explicit wrapping options whatever we do.)

The second process is heavy-handed. I am generally reluctant to abort an entire process as it is irrecoverable. However, this is only in debug mode -- but it impacts users who want to use checked overflowing even in optimized code, as aborting the process may not be an option for them.

Aborting the thread is insufficient in general since others may observe the values.

Another option is to wrap but return an enum that indicates if wrapping occurred. This seems to add some overhead and mainly serves to warn users that wrapping can occur, thus encouraging them to design a recovery strategy. Probably not very appealing.

Falling back to a CAS loop gives irregular performance similarly may impact users who wish to turn on checked overflow by default (even outside debug mode).

**Comment by nikomatsakis** [comment]
Having given the question of atomics some more thought (and talked it over with @aturon), my feeling is that the best option is simply to clearly document the semantics of `AtomicInt` and friends as wrapping on overflow. Naturally this is something of a tradeoff. As I see it, using atomic ints and the like is somewhat advanced, and always rather isolated: it is much more reasonable to expect users to be aware of the possibility of overflow in that scenario, as compared to general arithmetic throughout the program (atomics are also much easier to audit). Finally, given that the recovery options are typically either drastic or expensive, there isn't really an appealing alternative.

**Comment by pnkfelix** [comment]
@nikomatsakis maybe we should at least name them `AtomicWrappingInt` or something.  (but I don't know if they're even marked stable at the moment).

File: src/libcore/num/wrapping.rs (hunk 3 above)
Comment (oli-obk) [inline src/libcore/num/wrapping.rs]: floats don't panic on div by zero. they return inf.

Comment (nagisa) [inline src/libcore/num/wrapping.rs]: Err, I surely meant that integers’ division panics and never overflows when dividing.

**Comment by pnkfelix** [comment]
closing in favor of rebased version here; #22532

**Pull request closed without merging**